`lightning.tensor` mps docs (#756)
PennyLaneAI/pennylane-lightning: doc/lightning_tensor/device.rst

Lightning Tensor device
=======================

> The ``lightning.tensor`` device is a tensor network based device. It extends Pennylane-Lightning to run tensor network based simulations. 

Reviewer: ```suggestion
The ``lightning.tensor`` device is a tensor network simulator device.
```

(subjective, optional)

> The device is built on top of the `cutensornet` from the NVIDIA cuQuantum SDK, enabling GPU-accelerated simulation of quantum tensor network evolution.

Reviewer: ```suggestion
The device is built on top of the `cutensornet <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/index.html>`__ from the NVIDIA cuQuantum SDK, enabling GPU-accelerated simulation of quantum tensor network evolution.
```

If we dont want to link them, make it code with double ticks

A ``lightning.tensor`` device can be loaded using:

.. code-block:: python

    import pennylane as qml
    dev = qml.device("lightning.tensor", wires=100)

If the NVIDIA cuQuantum libraries are available, the above device will allow all operations to be performed on a CUDA capable GPU of generation SM 7.0 (Volta) and greater. 

> The ``lightning.tensor`` device also directly supports quantum circuit gradients using the ``parameter-shift`` method. This can be enabled at the PennyLane QNode level with:

Reviewer: Since parameter shift is the only viable option right now anyway, is this also the default or would a user have to manually set it?

Author: Users have to manually set `parameter-shift`. It's not a default option.

Reviewer: is it possible to make it the default or "best" for this device?

Author: Hey @Qottmann , it's technically fine to do that. But I would prefer to set a fast gradient method support method as `default` or `best`, probably like the `cutensornet`'s built-in gradient support (not there yet in ``cutensornet``) as `default`. ``parameter-shift`` is really slow.

.. code-block:: python

    @qml.qnode(dev, diff_method="parameter-shift")
    def circuit(params):
        ...

> Check out the :doc:`/lightning_tensor/installation` guide for more information.

Author: Suggested code does not work properly, so will keep previous implementation.

Lightning Tensor device
=======================

The ``lightning.tensor`` device is a tensor network simulator device. The device is built on top of the `cutensornet <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/index.html>`__ from the NVIDIA cuQuantum SDK, enabling GPU-accelerated simulation of quantum tensor network evolution.

A ``lightning.tensor`` device with the default setup can be simply loaded using:

.. code-block:: python

    import pennylane as qml
    dev = qml.device("lightning.tensor", wires=100)

By default, the device employs the ``Matrix Product State (MPS)`` tensor network approximation to represent the quantum state. 
Within this framework, the maximum bond dimension (``max_bond_dim``) defaults to ``128``. To manage computational complexity, 
singular value truncation is utilized, with the default threshold for truncation (``cutoff``) set at ``0``. Additionally, the 
singular value truncation mode (``cutoff_mode``) defaults to ``abs``, considering the absolute values of the singular values.

> Alternatively, users can opt to set ``cutoff_mode`` to ``rel`` to consider the relative values of the singular values.

Reviewer: The info provided is great! I would find it more accessible if they are listed in bullet points as (keyword) arguments as is typically done in code documentation. This is not a strict requirement, I just personally find walls of text in docs off-putting (subjective)

Lightning Tensor device
=======================

> The ``lightning.tensor`` device is a tensor network simulator device. The device is built on top of the `cutensornet <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/index.html>`__ from the NVIDIA cuQuantum SDK, enabling GPU-accelerated simulation of quantum tensor network evolution.

Reviewer 2: ```suggestion
The ``lightning.tensor`` device is a tensor network simulator device. The device is a frontend for the `cutensornet <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/index.html>`__ library from the NVIDIA cuQuantum SDK, enabling GPU-accelerated simulation of quantum tensor network evolution.
```

> A ``lightning.tensor`` device with the default setup can be simply loaded using:

Reviewer 2: ```suggestion
A ``lightning.tensor`` device can be loaded simply using:
```

.. code-block:: python

    import pennylane as qml
    dev = qml.device("lightning.tensor", wires=100)

> By default, the device employs the ``Matrix Product State (MPS)`` tensor network approximation to represent the quantum state. 

Reviewer 2: ```suggestion
By default, the device represents the quantum state approximately as a Matrix Product State (MPS). 
```

> Within this framework, the maximum bond dimension (``max_bond_dim``) defaults to ``128``. To manage computational complexity, 
> singular value truncation is utilized, with the default threshold for truncation (``cutoff``) set at ``0``. Additionally, the 

Reviewer 2: ```suggestion
Within this framework, the maximum bond dimension (``max_bond_dim``) defaults to ``128`` and the truncation cutoff (``cutoff``) to ``0``. Computational complexity is bounded satisfying either of the criteria, keeping the ``max_bond_dim`` largest singular values, or truncating the singular values smaller than ``cutoff``.
Additionally, the 
```

Author: reformated the context above with bullet points

singular value truncation mode (``cutoff_mode``) defaults to ``abs``, considering the absolute values of the singular values.
Alternatively, users can opt to set ``cutoff_mode`` to ``rel`` to consider the relative values of the singular values.

> The ``lightning.tensor`` device dispatches all operations to be performed on a CUDA capable GPU of generation SM 7.0 (Volta)

Reviewer 2: ```suggestion
The ``lightning.tensor`` device dispatches all operations to be performed on a CUDA-capable GPU of generation SM 7.0 (Volta)
```

and greater.

Note: Given the inherent parallelism of GPUs, simulations with intensive parallel computation, such as those with larger maximum

> bond dimensions, stand to gain the most from the computational power offered by GPU and those simulations can be benifited from the 

Reviewer 2: ```suggestion
bond dimensions, stand to gain the most from the computational power offered by GPU and those simulations can benefit from the 
```

Author: Good catch!

    ~pennylane.CY
    ~pennylane.CZ
    ~pennylane.DiagonalQubitUnitary
    ~pennylane.DoubleExcitation
    ~pennylane.ECR
    ~pennylane.Hadamard
    ~pennylane.Identity
    ~pennylane.IsingXX
    ~pennylane.IsingXY
    ~pennylane.IsingYY
    ~pennylane.IsingZZ
    ~pennylane.ISWAP
    ~pennylane.OrbitalRotation
    ~pennylane.PauliX
    ~pennylane.PauliY
    ~pennylane.PauliZ
    ~pennylane.PhaseShift
    ~pennylane.PSWAP
    ~pennylane.QFT
    ~pennylane.QubitCarry
    ~pennylane.QubitSum
    ~pennylane.QubitUnitary
    ~pennylane.Rot
    ~pennylane.RX
    ~pennylane.RY
    ~pennylane.RZ
    ~pennylane.S
    ~pennylane.SingleExcitation
    ~pennylane.SingleExcitationMinus
    ~pennylane.SingleExcitationPlus
    ~pennylane.SISWAP
    ~pennylane.SQISW
    ~pennylane.SWAP
    ~pennylane.SX
    ~pennylane.T
    ~pennylane.Toffoli

.. raw:: html

    </div>

**Unsupported operations:**

.. raw:: html

    <div class="summary-table">

.. autosummary::
    :nosignatures:

    ~pennylane.StatePrep
    ~pennylane.QubitStateVector
    ~pennylane.DoubleExcitationMinus
    ~pennylane.DoubleExcitationPlus
    ~pennylane.GlobalPhase

.. raw:: html

    </div>

**Supported observables:**

> The ``lightning.tensor`` supports all observables supported by lightning state-vector simulators, besides ``qml.SparseHamiltonian``, ``qml.Projector`` and limited support to ``qml.Hamiltonian``, ``qml.Prod``.

Reviewer 2: ```suggestion
The ``lightning.tensor`` supports all observables supported by the Lightning state-vector simulators, besides ``qml.SparseHamiltonian``, ``qml.Projector`` and limited support to ``qml.Hamiltonian``, ``qml.Prod``.
```

Lightning Tensor device
=======================

The ``lightning.tensor`` device is a tensor network simulator device. The device is built on top of the `cutensornet <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/index.html>`__ from the NVIDIA cuQuantum SDK, enabling GPU-accelerated simulation of quantum tensor network evolution.

A ``lightning.tensor`` device can be loaded simply using:

.. code-block:: python

    import pennylane as qml
    dev = qml.device("lightning.tensor", wires=100)

By default, the device represents the quantum state approximated as a Matrix Product State (MPS).

> The default setup for the MPS tensor network approximation are:

Reviewer 3: ```suggestion
The default setup for the MPS tensor network approximation is:
```

    ``lightning.tensor`` device.  It's worth noting that if the bond dimension used in the simulation is small, the ``lightning.tensor`` 
    device with ``MPS`` running a GPU may perform slower compared to a ``default.tensor`` device with ``MPS`` running on a CPU. For more details
    on how bond dimension affects the simulation performance, please refer to the ``Approximate Tensor Network Methods`` section in the `cuQuantum SDK <https://developer.nvidia.com/cuquantum-sdk>`__.

Users also have the flexibility to customize these parameters according to their specific needs with:

.. code-block:: python
    
    import pennylane as qml
    import numpy as np
    
    num_qubits = 100

    device_kwargs_mps = {
        "max_bond_dim": 64,
        "cutoff": 1e-10,
        "cutoff_mode": "abs",
    }

    dev = qml.device("lightning.tensor", wires=num_qubits, method="mps", **device_kwargs_mps)

The ``lightning.tensor`` device allows users to get quantum circuit gradients using the ``parameter-shift`` method. This can be enabled at the PennyLane ``QNode`` level with:

.. code-block:: python

    @qml.qnode(dev, diff_method="parameter-shift")
    def circuit(params):

>         ...

Reviewer 3: It would be great to have somewhere what is the default differentiation method.

Check out the :doc:`/lightning_tensor/installation` guide for more information.

> .. seealso:: :class:`~DefaultTensor` for a CPU only tensor network simulator device.

Reviewer 3: Should this be a link? It is not working in the docs.

Author: Nice catch! Thanks

Operations and observables support

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

> The ``lightning.tensor`` only supports 1,2-wires gates operations and gates operations that can be decomposed by PennyLane into 1,2-wires gates.  

Reviewer 3: ```suggestion
The "lightning.tensor" supports 1- and 2-wire gate operations and all other operations that can be decomposed to that by PennyLane.
```
